provider/oracle/environ.go

```diff
@@ -0,0 +1,731 @@
+	logger.Debugf("Getting instance networks")
+	networking, err := o.getInstanceNetworks(args, secLists, []string{vnicSets.Name})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	logger.Debugf("Getting cloud config")
+	cloudcfg, err := o.getCloudInitConfig(args.InstanceConfig.Series, networking)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Debugf("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OracleRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	logger.Debugf("oracle user data: %d bytes", len(userData))
+
+	attributes := map[string]interface{}{
+		"userdata": string(userData),
+	}
+
+	// create the instance based on the instance params
+	instance, err := o.createInstance(oci.InstanceParams{
+		Instances: []oci.Instances{
+			{
+				Shape:       spec.InstanceType.Name,
+				Imagelist:   imageName,
+				Name:        machineName,
+				Label:       args.InstanceConfig.MachineAgentServiceName,
+				Hostname:    args.InstanceConfig.MachineAgentServiceName,
+				Tags:        tags,
+				Attributes:  attributes,
+				Reverse_dns: false,
+				Networking:  networking,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance.arch = &spec.Image.Arch
+	instance.instType = &spec.InstanceType
+
+	machineId := instance.machine.Name
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(desiredStatus, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("started instance %q", machineId)
+	//TODO: add config option for public IP allocation
+	logger.Debugf("Associating public IP to instance %q", machineId)
+	if err := instance.associatePublicIP(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	result := &environs.StartInstanceResult{
+		Instance: instance,
+		Hardware: instance.hardwareCharacteristics(),
+	}
+
+	return result, nil
+}
+
+// StopInstances is part of the InstanceBroker interface.
+func (o *oracleEnviron) StopInstances(ids ...instance.Id) error {
+	oracleInstances, err := o.getOracleInstances(ids...)
+	if err == environs.ErrNoInstances {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logger.Debugf("terminating instances %v", ids)
+	if err := o.terminateInstances(oracleInstances...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *oracleEnviron) terminateInstances(instances ...*oracleInstance) error {
+	wg := sync.WaitGroup{}
+	errc := make(chan error, len(instances))
+	wg.Add(len(instances))
+	for _, oInst := range instances {
+		go func() {
+			defer wg.Done()
+			if err := oInst.delete(true); err != nil {
+				if !oci.IsNotFound(err) {
+					errc <- err
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	select {
+	case err := <-errc:
+		return errors.Annotate(err, "cannot delete all instances")
```

Comment:
This means the user won't see the errors for the other instances. I would prefer if you combined the errors if there is more than one.